### ISSUE TITLE
MiniProfiler 4 requires .netstandard2?

### ISSUE THREAD

**Issue by programcsharp** [MiniProfiler/dotnet#317]
I just tried to upgrade in an existing .NET 4.7.2 app and got a bunch of .netstandard2 dependencies in which caused conflicts. Had to roll back to v3.

Looks like Microsoft.Extensions.DependencyInjection.Abstractions only comes in a .netstandard2 version which is causing this.

Am I doing something wrong or is this by design?

**Comment by NickCraver** [comment]
Yep, that's how `netstandard` works - .NET 4.7.2 will have he actual DLLs on-box for most packages (System.*), but when transitive dependencies were added, a chain following install is fairly normal now.

What conflicts did you run into?

**Comment by programcsharp** [comment]
Well, I blew up production, so I rolled that back lickety split. Worked fine on my box, but once I pushed the build with the 50 different System.XXX.dll files for netstandard to the bin on prod, fireworks ensued -- dll cannot be loaded etc. So much for testing in production 👎 

If that package is required, I can try doing the upgrade again and push to a copy of the prod environment first and see what the basic issue is.

**Comment by NickCraver** [comment]
Some details that'd help:
- Which version of .NET are you on locally?
- Which version of .NET is on the server(s)?
- Which version of .NET is your project targeting?

**Comment by programcsharp** [comment]
OK, some details...

Local .NET version: 4.7.2
Server .NET version: 4.7.1
Project target framework: 4.7
Web.config targetFramework: 4.7.1

The exact exception is:

Method 'Match' in type 'TheApplication.WebApi.RegexOptionalConstraint' from assembly 'TheApplication, Version=2.29.0.36946, Culture=neutral, PublicKeyToken=null' does not have an implementation.

And that class is very simple, see below. Assuming there is some conflict between all of the different version shenanigans at play here. Any thoughts?

```
// http://chsakell.com/2013/10/13/web-api-custom-routing-constraints/
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Web.Http;
using System.Web.Http.Routing;

namespace TheApplication.WebApi
{
    public class RegexOptionalConstraint : IHttpRouteConstraint
    {
        readonly Regex _regEx;

        public RegexOptionalConstraint(string regEx)
        {
            if (string.IsNullOrWhiteSpace(regEx))
            {
                throw new ArgumentNullException("regEx");
            }

            _regEx = new Regex(regEx, RegexOptions.Compiled);
        }

        public bool Match(HttpRequestMessage request, IHttpRoute route, string parameterName,
            IDictionary<string, object> values, HttpRouteDirection routeDirection)
        {

            if (values[parameterName] != RouteParameter.Optional)
            {
                object value;

                if (values.TryGetValue(parameterName, out value))
                {
                    string input = Convert.ToString(value, CultureInfo.InvariantCulture);

                    return _regEx.IsMatch(input);
                }
                else
                    return false;
            }

            return true;
        }
    }
}
```

**Comment by NickCraver** [comment]
Unfortunately there are a few corner cases on `netstandard` which are a royal pain - as library authors we're still dealing with the fallout.

TL;DR on a fix is usually (and I think in this case): target higher when you can. Can you target net471 in your project and see what happens? I'm curious if that still has issues. net472 fixed *most* things, but net471 fixed a fair deal before that as well...hopefully we can get lucky here given your server versions.

**Comment by rubans** [comment]
I upgraded to .net 4.7.2 and still had the same error.
I can't believe v4 MiniProfiler has a dependency on .net standards 2 or have I misunderstood the issue?

**Comment by italodirenzo93** [comment]
Just want to add that I'm finding this to be caused by which version of Visual Studio you're using. Our application targets v4.7.2 but when attempting to upgrade the packages in VS2015, it fails. However, opening the same project in VS2017 and attempting to upgrade there succeeds. 🤷‍♂️

**Comment by NickCraver** [comment]
@italodirenzo93 If you're using `netstandard` *anything*, you'll need the latest NuGet in VS 2015. It's the package handler that's the issue there...not the individual packages.

In short, for all: this is a tooling issue. `netstandard` and the rules around it are relatively new (last few years), you'll need to make sure your NuGet tooling is up to date to handle it appropriately.

**Comment by NickCraver** [comment]
Going to close this out since the answer is: update your tooling and you'll be good to go :)

**Comment by programcsharp** [comment]
I looped around to this again and figured it out. For some reason, bad tooling or just because this project is a decade old and upgraded a million times, it wasn't adding the right references.

To fix web projects I added a reference in the web.config in system.web\compilation\assemblies:

```
    <compilation debug="true" targetFramework="4.7.2">
      <assemblies>
        <add assembly="netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" />
      </assemblies>
    </compilation>
```

To fix class library projects, I added a reference in the .csproj:

```
    <Reference Include="netstandard">
      <Private>True</Private>
    </Reference>
```

Hope this helps anybody else without the magic tooling 👿
